### DISCUSSION TITLE
Cannot add a custom locale via `config/locales` if PHP intl extension is installed.

### DISCUSSION THREAD

**Discussion by thisisjamessmith** [craftcms/cms#8924] [Ideas] [answered]
### Description

I want to add some custom locales that are not included in PHP's list generated by the intl extension. These reflect the languages available under Settings > Sites > [choose site] > Language. When the intl extension is not installed, you can add custom locales in `config/locales` and they successfully appear, but when it is installed, you cannot add to the ones generated by the extension.

### Steps to reproduce

1. Have an installation with PHP's intl extension installed
2. Add a custom locale file into `config/locales` that doesn't exist in PHP's intl list such as `en-ID.php` (English Indonesia)
3. Add this into general.php:

```
'extraAppLocales' => [
    'English (Indonesia)' => 'en-ID',
],
```

4. Get an error when loading the CP:
```Invalid language: en-id in craftcms/cms/src/helpers/Localization.php:44```

### Additional info

- Craft version: 3.4.28.1
- PHP version: 7.1

**Comment by brandonkelly** [comment]
Marked this as an enhancement request as this currently works by design. The Intl extension is supposed to provide an exhaustive list of recognized locales

**Comment by anvart** [comment]
I've got a problem with `Kurmanji` (Kursidh). Called `ResourceBundle::getLocales('')` to get list of locales and did not find it there. 

I also received `Invalid language` error. Are there any workarounds?

**Reply by brandonkelly** [reply]
@anvart Is that the same as `ku-TR`?

<img width="389" alt="A language input’s autocomplete list when searching for “Kurdish”." src="https://user-images.githubusercontent.com/47792/207179218-d5a3c2d6-c9cf-47ac-8e3f-7ecc45f1c58c.png">

**Reply by anvart** [reply]
@brandonkelly system thing, looks like. Even if I manage to add the locale locally, my hosting won't do it. Anyway, thank you for checking.

**Comment by bymayo** [comment]
I don't really understand why this is an option in Craft if it doesn't work?

Intl extension is installed and it doesn't work, and i also added a config file in to /config/locales/ and still get the same error:

```
An Error occurred while handling another error:
yii\base\InvalidArgumentException: Invalid language: zh-ch in /path/to/vendor/craftcms/cms/src/helpers/Localization.php:44
```

**Reply by brandonkelly** [reply]
`zh-ch` doesn’t seem to be a locale ID supported by ICU/Intl. `extraAppLocales` must be set to known locale IDs.

**Reply by angrybrad** [reply]
@bymayo `zh-ch` implies Chinese Switzerland, which would certainly be an odd language code. Did you mean `zh-cn` for Chinese Simplified?

**Reply by bymayo** [reply]
Sorry my bad, it was meant to be `zh-cn` 😂 How do I check if its one supported by ICU/Intl?

**Reply by angrybrad** [reply]
@bymayo not a straightforward way, but https://stackoverflow.com/questions/10073425/is-there-a-way-to-get-the-list-of-available-locales-in-php has some options.

The issue is Craft relies on PHP's intl extension for its localization data https://www.php.net/manual/en/book.intl.php

The intl extension relies on the ICU library installed on the server: https://unicode-org.github.io/icu/

The ICU library ultimately gets its data from the CLDR project: https://cldr.unicode.org/

So even if you have the intl extension installed, it’s probably using your distro’s default version of ICU, which may or may not have the language codes you’re looking for. If that’s the case, and assuming CLDR actually has the language code you need, you’d have to compile that database with the new info so it’s exposed to the intl extension.

**Comment by tibbis** [comment]
Checking in here to see if you can help me adding a non-standard locale? E.g. en-FR (english speakers in France). Tried using: 
`->extraAppLocales(['en-FR'])` in general.php but no luck. How can I add this? 

(Use case: for hreflang links between regions)

**Comment by brandonkelly** [comment, marked as answer]
In Craft 5, you can add custom locales using the [localeAliases](https://craftcms.com/docs/5.x/reference/config/general.html#localealiases) config setting:

```php
->localeAliases([
    'en-FR' => [
        'aliasOf' => 'en',
        'displayName' => 'English (French)',
    ],
]);
```

**Reply by tibbis** [reply]
ah missed this! thanks
